src/quic/streams.cc

```diff
@@ -1,61 +1,1052 @@
+  V(SetPriority, setPriority, false)                                           \
+  V(GetPriority, getPriority, true)                                            \
+  V(GetReader, getReader, false)
+
+struct Stream::State {
+#define V(_, name, type) type name;
+  STREAM_STATE(V)
+#undef V
+};
+
+STAT_STRUCT(Stream, STREAM)
+
+// ============================================================================
+
+namespace {
+Maybe<std::shared_ptr<DataQueue>> GetDataQueueFromSource(Environment* env,
+                                                         Local<Value> value) {
+  DCHECK_IMPLIES(!value->IsUndefined(), value->IsObject());
+  if (value->IsUndefined()) {
+    return Just(std::shared_ptr<DataQueue>());
+  } else if (value->IsArrayBuffer()) {
+    auto buffer = value.As<ArrayBuffer>();
+    std::vector<std::unique_ptr<DataQueue::Entry>> entries(1);
+    entries.push_back(DataQueue::CreateInMemoryEntryFromBackingStore(
+        buffer->GetBackingStore(), 0, buffer->ByteLength()));
+    return Just(DataQueue::CreateIdempotent(std::move(entries)));
+  } else if (value->IsSharedArrayBuffer()) {
+    auto buffer = value.As<SharedArrayBuffer>();
+    std::vector<std::unique_ptr<DataQueue::Entry>> entries(1);
+    entries.push_back(DataQueue::CreateInMemoryEntryFromBackingStore(
+        buffer->GetBackingStore(), 0, buffer->ByteLength()));
+    return Just(DataQueue::CreateIdempotent(std::move(entries)));
+  } else if (value->IsArrayBufferView()) {
+    std::vector<std::unique_ptr<DataQueue::Entry>> entries(1);
+    entries.push_back(
+        DataQueue::CreateInMemoryEntryFromView(value.As<ArrayBufferView>()));
+    return Just(DataQueue::CreateIdempotent(std::move(entries)));
+  } else if (Blob::HasInstance(env, value)) {
+    Blob* blob;
+    ASSIGN_OR_RETURN_UNWRAP(
+        &blob, value, Nothing<std::shared_ptr<DataQueue>>());
+    return Just(blob->getDataQueue().slice(0));
+  }
+  // TODO(jasnell): Add streaming sources...
+  THROW_ERR_INVALID_ARG_TYPE(env, "Invalid data source type");
+  return Nothing<std::shared_ptr<DataQueue>>();
+}
+}  // namespace
+
+struct Stream::Impl {
+  static void AttachSource(const FunctionCallbackInfo<Value>& args) {
+    Environment* env = Environment::GetCurrent(args);
+
+    std::shared_ptr<DataQueue> dataqueue;
+    if (GetDataQueueFromSource(env, args[0]).To(&dataqueue)) {
+      Stream* stream;
+      ASSIGN_OR_RETURN_UNWRAP(&stream, args.Holder());
+      stream->set_outbound(std::move(dataqueue));
+    }
+  }
+
+  static void Destroy(const FunctionCallbackInfo<Value>& args) {
+    Stream* stream;
+    ASSIGN_OR_RETURN_UNWRAP(&stream, args.Holder());
+    if (args.Length() > 1) {
+      CHECK(args[0]->IsBigInt());
+      bool unused = false;
+      stream->Destroy(QuicError::ForApplication(
+          args[0].As<BigInt>()->Uint64Value(&unused)));
+    } else {
+      stream->Destroy();
+    }
+  }
+
+  static void SendHeaders(const FunctionCallbackInfo<Value>& args) {
```

Comment:
Since it returns bool and all parameters are supported, we can add Fast API here. Can you add a comment here? 

```
 * Currently supported argument types:
 *  - pointer to an embedder type
 *  - JavaScript array of primitive types
 *  - bool
 *  - int32_t
 *  - uint32_t
 *  - int64_t
 *  - uint64_t
 *  - float32_t
 *  - float64_t
```

Reply:
I'll be going back and doing fast api variations after the main implementation is done.